src/controller.cpp

```diff
@@ -1019,36 +1046,53 @@ bool Controller::activate(const ServerData& serverData,
                 return;
               }
 
-              setState(StateConnecting);
+              if (initiator != ExtensionUser) {
+                setState(StateConnecting);
+              }
 
               clearRetryCounter();
-              activateInternal(DoNotForceDNSPort, serverSelectionPolicy);
+              activateInternal(DoNotForceDNSPort, serverSelectionPolicy,
+                               initiator);
             });
 
     connect(request, &NetworkRequest::requestCompleted, this,
-            [this, serverSelectionPolicy](const QByteArray& data) {
+            [this, serverSelectionPolicy, initiator](const QByteArray& data) {
               MozillaVPN::instance()->accountChecked(data);
-              setState(StateConnecting);
+
+              if (initiator != ExtensionUser) {
+                setState(StateConnecting);
+              }
 
               clearRetryCounter();
-              activateInternal(DoNotForceDNSPort, serverSelectionPolicy);
+              activateInternal(DoNotForceDNSPort, serverSelectionPolicy,
+                               initiator);
             });
 
     connect(request, &QObject::destroyed, task, &QObject::deleteLater);
     return true;
   }
+  if (initiator != ExtensionUser) {
+    setState(StateConnecting);
+  }
 
   clearRetryCounter();
-  activateInternal(DoNotForceDNSPort, serverSelectionPolicy);
+  activateInternal(DoNotForceDNSPort, serverSelectionPolicy, initiator);
   return true;
 }
 
-bool Controller::deactivate() {
+bool Controller::deactivate(ActivationPrincipal user) {
   logger.debug() << "Deactivation" << m_state;
+  if (m_initiator > user) {
```

Comment:
Eventually we want to embellish this to allow deactivating the VPN _for Firefox_ from the extension in this situation though right?

Reply:
probably not. :) 
A: We cannot know that whatever is talking using the web-extension api really is firefox. 
B: We don't need to disable the vpn for firefox to not use the vpn connection see https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9676